local-gpio.c

```diff
@@ -0,0 +1,329 @@
+ *   - polarity: ACTIVE_HIGH or ACTIVE_LOW
+ *
+ * Example: gpiochip0;POWER,0,ACTIVE_LOW;FASTBOOT_KEY,1,ACTIVE_HIGH;POWER_KEY,2,ACTIVE_HIGH;USB_DISCONNECT,3,ACTIVE_LOW
+ */
+
+static void local_gpio_parse_config(struct local_gpio *local_gpio, char *control_dev)
+{
+	char *c;
+	size_t device_len;
+
+	// First liblocal description
+	c = strchr(control_dev, ';');
+	if (!c)
+		device_len = strlen(control_dev);
+	else
+		device_len = c - control_dev;
+
+	local_gpio->gpiochip_desc = strndup(control_dev, device_len);
+
+	if (!c)
+		return;
+
+	// GPIOs
+	while(c) {
+		char *name, *off, *pol;
+		unsigned gpio_type;
+		unsigned gpio_offset;
+		unsigned gpio_polarity;
+
+		name = c + 1;
+		off = strchr(name, ',');
+		if (!off)
+			errx(1, "GPIOs config invalid");
+		off += 1;
+		pol = strchr(off, ',');
+		if (!pol)
+			errx(1, "GPIOs config invalid");
+		pol += 1;
+
+		c = strchr(pol, ';');
+
+		if (strncmp("POWER", name, off - name - 1) == 0)
+			gpio_type = GPIO_POWER;
+		else if (strncmp("FASTBOOT_KEY", name, off - name - 1) == 0)
+			gpio_type = GPIO_FASTBOOT_KEY;
+		else if (strncmp("POWER_KEY", name, off - name - 1) == 0)
+			gpio_type = GPIO_POWER_KEY;
+		else if (strncmp("USB_DISCONNECT", name, off - name - 1) == 0)
+			gpio_type = GPIO_USB_DISCONNECT;
+		else
+			errx(1, "GPIOs type invalid: '%s'", name);
+
+		gpio_offset = strtoul(off, NULL, 0);
+
+		if (strncmp("ACTIVE_HIGH", pol, c - pol - 1) == 0)
+			gpio_polarity = GPIO_ACTIVE_HIGH;
+		else if (strncmp("ACTIVE_LOW", pol, c - pol - 1) == 0)
+			gpio_polarity = GPIO_ACTIVE_LOW;
+		else
+			errx(1, "GPIOs polarity invalid: '%s'", pol);
+
+		local_gpio->gpio_present[gpio_type] = 1;
+		local_gpio->gpio_offset[gpio_type] = gpio_offset;
+		local_gpio->gpio_polarity[gpio_type] = gpio_polarity;
+	}
+}
+
+void *local_gpio_open(struct device *dev)
+{
+#if GPIOD_MAJOR_VERSION == 2
+	struct gpiod_request_config *req_cfg;
+	char *gpiochip_path;
+#endif
+	struct local_gpio *local_gpio;
+	int i;
+
+	local_gpio = calloc(1, sizeof(*local_gpio));
+
+	local_gpio_parse_config(local_gpio, dev->control_dev);
+
+#if GPIOD_MAJOR_VERSION == 2
+	if (asprintf(&gpiochip_path, "/dev/%s", local_gpio->gpiochip_desc) < 0) {
+		free(local_gpio);
+		return NULL;
+	}
+
+	local_gpio->chip = gpiod_chip_open(gpiochip_path);
+#else
+	local_gpio->chip = gpiod_chip_open_lookup(local_gpio->gpiochip_desc);
+#endif
+	if (!local_gpio->chip) {
+		err(1, "Unable to open gpiochip '%s'", local_gpio->gpiochip_desc);
+		return NULL;
+	}
+
+#if GPIOD_MAJOR_VERSION == 2
+	req_cfg = gpiod_request_config_new();
+	if (!req_cfg) {
+		err(1, "Unable to allocate request config");
+		return NULL;
+	}
+	gpiod_request_config_set_consumer(req_cfg, "cdba");
+#endif
+
+	for (i = 0; i < GPIO_COUNT; ++i) {
+#if GPIOD_MAJOR_VERSION == 2
+		struct gpiod_line_settings *line_settings;
+		struct gpiod_line_config *line_cfg;
+#else
+		struct gpiod_line_request_config cfg;
+#endif
+
+		if (!local_gpio->gpio_present[i])
+			continue;
+
+#if GPIOD_MAJOR_VERSION == 2
```

Comment:
This is complete ugh. I'd say, either drop support for older libgpiod, or provide a wrapping interface.

Reply:
Blame @brgl for that since both versions are totally incompatible, and v2 is shipped only with latest fedora, but v1 is shipped with all the other distros so far:
https://packages.fedoraproject.org/pkgs/libgpiod/libgpiod/
https://packages.debian.org/search?keywords=libgpiod&searchon=names&suite=stable&section=all
https://packages.ubuntu.com/search?keywords=libgpiod&searchon=names&suite=kinetic&section=all
https://pkgs.alpinelinux.org/packages?name=libgpiod&branch=edge&repo=&arch=&maintainer=
https://archlinux.org/packages/?sort=&q=gpiod&maintainer=&flagged=
...

Can you explicit what is a "wrapping interface" for you ? a separate c file for each major version ?

Reply:
Well v2 is built around the kernel v2 uAPI for GPIO - which was a complete re-write of the interface. I couldn't make it v1 compatible if I tried. I make sure yocto supports the latest release upstream but I can't possibly make all distros switch to v2. :(

Reply:
As I wrote, a wrapper code can help.

Reply:
Regarding the 'wrapping interface'. I'd say, you can either implement one interface in terms of another one (of your choice), implement separate C files for v1 vs v2 implementing GPIO support or just simply two drivers and a shared common parsing function.

Reply:
Ack, I'll try different solutions

Reply:
There's one more solution: do switch to libgpiod and get it as a git submodule.

Reply:
Yep. I think this is called meson subproject or something suchalike. Meson can fetch external dependencies not present on the host system.

Reply:
I pushed an implementation this, the simplest and cleanest, now I need to work on the yaml part